File: src/content/library.rs
use super::snippets::BuiltinCodeSnippet;
use crate::model::EverydaySentenceLength;
use anyhow::{Context, Result};
use std::collections::HashSet;

const USER_EVERYDAY_CORPUS_ENV: &str = "KEYLOOP_EVERYDAY_CORPUS";

#[derive(Debug, Clone)]
pub struct ContentLibrary {
    pub warmup: Vec<String>,
    pub foundation_drills: Vec<FoundationDrill>,
    pub word_chunks: Vec<String>,
    pub common_words: Vec<String>,
    pub everyday_english: EverydayEnglishCorpus,
    pub programming_words: Vec<String>,
    pub symbols: Vec<String>,
    pub language_symbols: Vec<LanguageSymbolSet>,
    pub number_drills: Vec<String>,
    pub naming: Vec<String>,
    pub code_snippets: Vec<BuiltinCodeSnippet>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct FoundationDrill {
    pub id: String,
    pub title_zh: String,
    pub title_en: String,
    pub hint_zh: String,
    pub hint_en: String,
    pub items: Vec<String>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct LanguageSymbolSet {
    pub language: Option<String>,
    pub framework: Option<String>,
    pub items: Vec<String>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct SourceCatalogEntry {
    pub source_id: String,
    pub repo: String,
    pub repo_url: String,
    pub license_spdx: String,
    pub retrieved_at: String,
    pub languages: Vec<String>,
    pub frameworks: Vec<String>,
    pub notes: String,
    #[serde(default)]
    pub source_name: String,
    #[serde(default)]
    pub source_url: String,
    #[serde(default)]
    pub corpus: String,
    #[serde(default)]
    pub generation_script: String,
    #[serde(default)]
    pub included_fields: Vec<String>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct EverydayEnglishCorpus {
    pub sources: Vec<EverydayCorpusSource>,
    pub entries: Vec<EverydayCorpusEntry>,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct EverydayCorpusSource {
    pub source_id: String,
    pub source_name: String,
    pub source_url: String,
    pub license: String,
    pub retrieved_at: String,
    pub generation_script: String,
    pub included_fields: Vec<String>,
    pub notes: String,
}

#[derive(Debug, Clone, serde::Deserialize)]
pub struct EverydayCorpusEntry {
    pub text: String,
    pub kind: EverydayEntryKind,
    pub tier: Option<u8>,
    pub length: Option<EverydaySentenceLength>,
    pub domain: EverydayEntryDomain,
    pub source_id: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EverydayEntryKind {
    Word,
    Phrase,
    Sentence,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EverydayEntryDomain {
    Everyday,
    Workplace,
}

pub fn load() -> Result<ContentLibrary> {
    let everyday_english = load_everyday_english_corpus()?;

    Ok(ContentLibrary {
        warmup: load_json_list("warmup", include_str!("../../content/warmup.json"))?,
        foundation_drills: load_json_records(
            "foundation_drills",
            include_str!("../../content/foundation_drills.json"),
        )?,
        word_chunks: load_json_list(
            "word_chunks",
            include_str!("../../content/word_chunks.json"),
        )?,
        common_words: load_json_list(
            "common_words",
            include_str!("../../content/common_words.json"),
        )?,
        everyday_english,
        programming_words: load_json_list(
            "programming_words",
            include_str!("../../content/programming_words.json"),
        )?,
        symbols: load_json_list("symbols", include_str!("../../content/symbols.json"))?,
        language_symbols: load_json_records(
            "language_symbols",
            include_str!("../../content/language_symbols.json"),
        )?,
        number_drills: load_json_list(
            "number_drills",
            include_str!("../../content/number_drills.json"),
        )?,
        naming: load_json_list("naming", include_str!("../../content/naming.json"))?,
        code_snippets: load_code_snippets()?,
    })
}

pub fn source_catalog() -> Result<Vec<SourceCatalogEntry>> {
    let mut sources = load_json_records::<SourceCatalogEntry>(
        "source_catalog",
        include_str!("../../content/source_catalog.json"),
    )?;
    let everyday = load_everyday_english_corpus()?;
    sources.extend(everyday.sources.into_iter().map(SourceCatalogEntry::from));
    Ok(sources)
}

impl From<EverydayCorpusSource> for SourceCatalogEntry {
    fn from(source: EverydayCorpusSource) -> Self {
        Self {
            source_id: source.source_id,
            repo: source.source_name.clone(),
            repo_url: source.source_url.clone(),
            license_spdx: source.license,
            retrieved_at: source.retrieved_at,
            languages: vec!["english".to_string()],
            frameworks: vec!["everyday".to_string(), "workplace".to_string()],
            notes: source.notes,
            source_name: source.source_name,
            source_url: source.source_url,
            corpus: "everyday_english".to_string(),
            generation_script: source.generation_script,
            included_fields: source.included_fields,
        }
    }
}

fn load_json_list(name: &str, data: &str) -> Result<Vec<String>> {
    serde_json::from_str(data).with_context(|| format!("Could not load content/{name}.json"))
}

fn load_code_snippets() -> Result<Vec<BuiltinCodeSnippet>> {
    let mut snippets = Vec::new();
    for (name, data) in [
        ("code/react", include_str!("../../content/code/react.json")),
        ("code/vue", include_str!("../../content/code/vue.json")),
        (
            "code/nestjs",
            include_str!("../../content/code/nestjs.json"),
        ),
        (
            "code/solidity",
            include_str!("../../content/code/solidity.json"),
        ),
        ("code/rust", include_str!("../../content/code/rust.json")),
        ("code/web", include_str!("../../content/code/web.json")),
        ("code/css", include_str!("../../content/code/css.json")),
        (
            "code/generated/typescript",
            include_str!("../../content/code/generated/typescript.json"),
        ),
        (
            "code/generated/javascript",
            include_str!("../../content/code/generated/javascript.json"),
        ),
        (
            "code/generated/vue",
            include_str!("../../content/code/generated/vue.json"),
        ),
        (
            "code/generated/solidity",
            include_str!("../../content/code/generated/solidity.json"),
        ),
        (
            "code/generated/rust",
            include_str!("../../content/code/generated/rust.json"),
        ),
        (
            "code/generated/html",
            include_str!("../../content/code/generated/html.json"),
        ),
        (
            "code/generated/css",
            include_str!("../../content/code/generated/css.json"),
        ),
        (
            "code/generated/scss",
            include_str!("../../content/code/generated/scss.json"),
        ),
        (
            "code/generated/less",
            include_str!("../../content/code/generated/less.json"),
        ),
    ] {
        snippets.extend(load_json_records::<BuiltinCodeSnippet>(name, data)?);
    }
    Ok(snippets)
}

fn load_everyday_english_corpus() -> Result<EverydayEnglishCorpus> {
    let mut corpus = load_json_value::<EverydayEnglishCorpus>(
        "everyday_english",
        include_str!("../../content/everyday_english.json"),
    )?;
    if let Some(user_corpus) = load_user_everyday_english_corpus()? {
        merge_everyday_corpus(&mut corpus, user_corpus);
    }
    Ok(corpus)
}

fn load_user_everyday_english_corpus() -> Result<Option<EverydayEnglishCorpus>> {
    let Ok(path) = std::env::var(USER_EVERYDAY_CORPUS_ENV) else {
        return Ok(None);
    };
    let path = path.trim();
    if path.is_empty() {
        return Ok(None);
    }

    let data = std::fs::read_to_string(path)
        .with_context(|| format!("Could not read {USER_EVERYDAY_CORPUS_ENV}={path}"))?;
    let corpus = serde_json::from_str(&data)
        .with_context(|| format!("Could not load {USER_EVERYDAY_CORPUS_ENV}={path}"))?;
    Ok(Some(corpus))
}

pub(crate) fn merge_everyday_corpus(
    base: &mut EverydayEnglishCorpus,
    extra: EverydayEnglishCorpus,
) {
    let mut source_ids = base
        .sources
        .iter()
        .map(|source| source.source_id.clone())
        .collect::<HashSet<_>>();
    for source in extra.sources {
        if source_ids.insert(source.source_id.clone()) {
            base.sources.push(source);
        }
    }

    let mut entry_keys = base
        .entries
        .iter()
        .map(|entry| (entry.source_id.clone(), entry.kind, entry.text.clone()))
        .collect::<HashSet<_>>();
    for entry in extra.entries {
        let key = (entry.source_id.clone(), entry.kind, entry.text.clone());
        if entry_keys.insert(key) {
            base.entries.push(entry);
        }
    }
}

fn load_json_records<T>(name: &str, data: &str) -> Result<Vec<T>>
where
    T: serde::de::DeserializeOwned,
{
    serde_json::from_str(data).with_context(|| format!("Could not load content/{name}.json"))
}

fn load_json_value<T>(name: &str, data: &str) -> Result<T>
where
    T: serde::de::DeserializeOwned,
{
    serde_json::from_str(data).with_context(|| format!("Could not load content/{name}.json"))
}

File: src/content/library.ts
import { access, readFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import type {
  EverydayLevel,
  EverydaySentenceLength,
  EverydayWordRange,
} from "../domain/model";
import type { CodePracticeOption } from "../domain/model";
import { codePracticeOptions, type BuiltinCodeSnippet } from "./snippets";
import type { LongWordEntry } from "../training/vocabulary";
import {
  codeCorpusPracticeOptions,
  loadCodeCorpus,
  type CodeCorpus,
} from "./codeCorpus";

export interface ContentLibrary {
  warmup: string[];
  foundation_drills: FoundationDrill[];
  word_chunks: string[];
  common_words: string[];
  everyday_english: EverydayEnglishCorpus;
  everyday_words: EverydayWordsCorpus;
  everyday_sentences: EverydaySentencesCorpus;
  everyday_articles: EverydayArticlesCorpus;
  everyday_word_decomposition: EverydayWordDecompositionCorpus;
  programming_words: ProgrammingWordEntry[];
  code_corpus?: CodeCorpus;
  code_snippets: BuiltinCodeSnippet[];
  long_words: LongWordEntry[];
}

export interface ProgrammingWordEntry {
  word: string;
  note_zh: string;
}

export interface FoundationDrill {
  id: string;
  title_zh: string;
  title_en: string;
  hint_zh: string;
  hint_en: string;
  items: string[];
}

export interface SourceCatalogEntry {
  source_id: string;
  repo: string;
  repo_url: string;
  license_spdx: string;
  retrieved_at: string;
  languages: string[];
  frameworks: string[];
  notes: string;
  source_name: string;
  source_url: string;
  corpus: string;
  generation_script: string;
  included_fields: string[];
}

export interface EverydayEnglishCorpus {
  sources: EverydayCorpusSource[];
  entries: EverydayCorpusEntry[];
}

export interface EverydayCorpusSource {
  source_id: string;
  source_name: string;
  source_url: string;
  license: string;
  retrieved_at: string;
  generation_script: string;
  included_fields: string[];
  notes: string;
}

export interface EverydayCorpusEntry {
  text: string;
  kind: "word" | "phrase" | "sentence";
  tier: number | null;
  length: EverydaySentenceLength | null;
  domain: "everyday" | "workplace";
  source_id: string;
}

export interface EverydayWordsCorpus {
  sources: EverydayCorpusSource[];
  entries: EverydayWordEntry[];
}

export interface EverydayWordEntry {
  word: string;
  rank: number;
  range: EverydayWordRange;
  level: EverydayLevel;
  translation_zh: string;
  source_id: string;
}

export interface EverydaySentencesCorpus {
  sources: EverydayCorpusSource[];
  entries: EverydaySentenceEntry[];
}

export interface EverydaySentenceEntry {
  text: string;
  translation_zh: string;
  level: EverydayLevel;
  length: EverydaySentenceLength;
  source_id: string;
  source_title: string;
}

export interface EverydayArticlesCorpus {
  sources: EverydayCorpusSource[];
  entries: EverydayArticleEntry[];
}

export interface EverydayArticleEntry {
  title: string;
  level: EverydayLevel;
  length: EverydaySentenceLength;
  source_id: string;
  paragraphs: EverydayArticleParagraph[];
}

export interface EverydayArticleParagraph {
  text: string;
  translation_zh: string;
}

export interface EverydayWordDecompositionCorpus {
  sources: EverydayCorpusSource[];
  entries: EverydayWordDecompositionEntry[];
}

export interface EverydayWordDecompositionEntry {
  word: string;
  parts: string[];
  translation_zh: string;
  level: EverydayLevel;
  source_id: string;
}

const contentMarkerFile = "everyday_english.json";
export async function loadContentLibrary(options: {
  userEverydayCorpusPath?: string;
} = {}): Promise<ContentLibrary> {
  const everydayEnglish = await loadEverydayEnglishCorpus(options.userEverydayCorpusPath);
  const codeCorpus = loadCodeCorpus();

  return {
    warmup: await loadJsonFile<string[]>("warmup.json"),
    foundation_drills: await loadJsonFile<FoundationDrill[]>("foundation_drills.json"),
    word_chunks: await loadJsonFile<string[]>("word_chunks.json"),
    common_words: await loadJsonFile<string[]>("common_words.json"),
    everyday_english: everydayEnglish,
    everyday_words: await loadJsonFile<EverydayWordsCorpus>("everyday_words.json"),
    everyday_sentences: await loadJsonFile<EverydaySentencesCorpus>("everyday_sentences.json"),
    everyday_articles: await loadJsonFile<EverydayArticlesCorpus>("everyday_articles.json"),
    everyday_word_decomposition: await loadJsonFile<EverydayWordDecompositionCorpus>(
      "everyday_word_decomposition.json",
    ),
    programming_words: await loadJsonFile<ProgrammingWordEntry[]>("programming_words.json"),
    code_corpus: codeCorpus,
    code_snippets: [],
    long_words: await loadJsonFile<LongWordEntry[]>("long_words.json"),
  };
}

export function codePracticeOptionsForLibrary(
  library: ContentLibrary,
): CodePracticeOption[] {
  if (library.code_corpus !== undefined) {
    return codeCorpusPracticeOptions(library.code_corpus);
  }
  return codePracticeOptions(library.code_snippets);
}

export async function sourceCatalog(options: {
  userEverydayCorpusPath?: string;
} = {}): Promise<SourceCatalogEntry[]> {
  const sources = await loadJsonFile<SourceCatalogEntry[]>("source_catalog.json");
  const everyday = await loadEverydayEnglishCorpus(options.userEverydayCorpusPath);
  const everydayWords = await loadJsonFile<EverydayWordsCorpus>("everyday_words.json");
  const everydaySentences = await loadJsonFile<EverydaySentencesCorpus>(
    "everyday_sentences.json",
  );
  const everydayArticles = await loadJsonFile<EverydayArticlesCorpus>(
    "everyday_articles.json",
  );
  const everydayWordDecomposition =
    await loadJsonFile<EverydayWordDecompositionCorpus>(
      "everyday_word_decomposition.json",
    );
  return [
    ...sources.map(normalizeSourceCatalogEntry),
    ...everyday.sources.map((source) =>
      everydaySourceToCatalogEntry(source, "everyday_english"),
    ),
    ...everydayWords.sources.map((source) =>
      everydaySourceToCatalogEntry(source, "everyday_words"),
    ),
    ...everydaySentences.sources.map((source) =>
      everydaySourceToCatalogEntry(source, "everyday_sentences"),
    ),
    ...everydayArticles.sources.map((source) =>
      everydaySourceToCatalogEntry(source, "everyday_articles"),
    ),
    ...everydayWordDecomposition.sources.map((source) =>
      everydaySourceToCatalogEntry(source, "everyday_word_decomposition"),
    ),
  ];
}

export interface ResolveContentRootOptions {
  moduleUrl?: string;
  execPath?: string;
  argv1?: string;
  env?: Record<string, string | undefined>;
  exists?: (path: string) => Promise<boolean>;
}

export async function resolveContentRoot(
  options: ResolveContentRootOptions = {},
): Promise<string> {
  const candidates = contentRootCandidates(options);
  const exists = options.exists ?? contentRootExists;
  for (const candidate of candidates) {
    if (await exists(candidate)) {
      return candidate;
    }
  }
  return candidates[0] ?? join(process.cwd(), "contents");
}

export function mergeEverydayCorpus(
  base: EverydayEnglishCorpus,
  extra: EverydayEnglishCorpus,
): EverydayEnglishCorpus {
  const sources = [...base.sources];
  const sourceIds = new Set(sources.map((source) => source.source_id));
  for (const source of extra.sources) {
    if (!sourceIds.has(source.source_id)) {
      sourceIds.add(source.source_id);
      sources.push(source);
    }
  }

  const entries = [...base.entries];
  const entryKeys = new Set(
    entries.map((entry) => `${entry.source_id}\u0000${entry.kind}\u0000${entry.text}`),
  );
  for (const entry of extra.entries) {
    const key = `${entry.source_id}\u0000${entry.kind}\u0000${entry.text}`;
    if (!entryKeys.has(key)) {
      entryKeys.add(key);
      entries.push(entry);
    }
  }

  return { sources, entries };
}

async function loadEverydayEnglishCorpus(
  userEverydayCorpusPath?: string,
): Promise<EverydayEnglishCorpus> {
  let corpus = await loadJsonFile<EverydayEnglishCorpus>("everyday_english.json");
  const path = userEverydayCorpusPath ?? process.env.KEYLOOP_EVERYDAY_CORPUS?.trim();
  if (path !== undefined && path.length > 0) {
    corpus = mergeEverydayCorpus(corpus, await loadJsonPath<EverydayEnglishCorpus>(path));
  }
  return corpus;
}

function everydaySourceToCatalogEntry(
  source: EverydayCorpusSource,
  corpus: string,
): SourceCatalogEntry {
  return {
    source_id: source.source_id,
    repo: source.source_name,
    repo_url: source.source_url,
    license_spdx: source.license,
    retrieved_at: source.retrieved_at,
    languages: ["english"],
    frameworks: ["everyday", "workplace"],
    notes: source.notes,
    source_name: source.source_name,
    source_url: source.source_url,
    corpus,
    generation_script: source.generation_script,
    included_fields: source.included_fields,
  };
}

function normalizeSourceCatalogEntry(source: SourceCatalogEntry): SourceCatalogEntry {
  return {
    source_id: source.source_id,
    repo: source.repo,
    repo_url: source.repo_url,
    license_spdx: source.license_spdx,
    retrieved_at: source.retrieved_at,
    languages: source.languages,
    frameworks: source.frameworks,
    notes: source.notes,
    source_name: source.source_name ?? "",
    source_url: source.source_url ?? "",
    corpus: source.corpus ?? "",
    generation_script: source.generation_script ?? "",
    included_fields: source.included_fields ?? [],
  };
}

async function loadJsonFile<T>(relativePath: string): Promise<T> {
  const contentRoot = await resolveContentRoot();
  return loadJsonPath<T>(join(contentRoot, relativePath));
}

async function loadJsonPath<T>(path: string): Promise<T> {
  return JSON.parse(await readFile(path, "utf8")) as T;
}

function contentRootCandidates(options: ResolveContentRootOptions): string[] {
  const env = options.env ?? process.env;
  const candidates: string[] = [];
  const tsEnvRoot = env.KEYLOOP_TS_CONTENT_ROOT?.trim();
  if (tsEnvRoot !== undefined && tsEnvRoot.length > 0) {
    candidates.push(resolve(tsEnvRoot));
  }
  const envRoot = env.KEYLOOP_CONTENT_ROOT?.trim();
  if (envRoot !== undefined && envRoot.length > 0) {
    candidates.push(resolve(envRoot));
  }

  const moduleUrl = options.moduleUrl ?? import.meta.url;
  try {
    candidates.push(
      join(
        dirname(dirname(dirname(fileURLToPath(moduleUrl)))),
        "contents",
      ),
    );
  } catch {
    // Non-file module URLs cannot identify the repository root.
  }

  addPathAdjacentContentCandidates(candidates, options.argv1 ?? process.argv[1]);
  addPathAdjacentContentCandidates(candidates, options.execPath ?? process.execPath);
  candidates.push(join(process.cwd(), "contents"));

  return [...new Set(candidates)];
}

function addPathAdjacentContentCandidates(
  candidates: string[],
  path: string | undefined,
): void {
  if (path === undefined || path.length === 0) {
    return;
  }
  const base = dirname(resolve(path));
  candidates.push(join(base, "contents"));
  candidates.push(join(base, "..", "contents"));
  candidates.push(join(base, "..", "..", "contents"));
  candidates.push(join(base, "..", "..", "..", "contents"));
}

async function contentRootExists(path: string): Promise<boolean> {
  try {
    await access(join(path, contentMarkerFile));
    return true;
  } catch {
    return false;
  }
}
